src/Workspaces/Core/Portable/Utilities/ValuesSources/ValueSource.cs

```diff
@@ -10,18 +13,12 @@ namespace Roslyn.Utilities
     /// </summary>
     internal abstract class ValueSource<T>
     {
-        public abstract bool TryGetValue(out T value);
+        public abstract bool TryGetValue([MaybeNullWhen(false)]out T value);
         public abstract T GetValue(CancellationToken cancellationToken = default);
         public abstract Task<T> GetValueAsync(CancellationToken cancellationToken = default);
 
-        public bool HasValue
-        {
-            get
-            {
-                return this.TryGetValue(out var tmp);
-            }
-        }
+        public bool HasValue => TryGetValue(out _);
 
-        public static readonly ConstantValueSource<T> Empty = new ConstantValueSource<T>(default);
+        public static readonly ConstantValueSource<T> Empty = new ConstantValueSource<T>(default!);
```

Comment:
💡 Need to file a follow-up issue to get an analyzer added ensuring that `ValueSource<T>.Empty` is only used when `T` is defaultable (either a value type, or a nullable reference type).

Reply:
How about

```C#
    internal static class ValueSource
    {
        public static ConstantValueSource<T> Empty<T>() where T : notnull
            => new ConstantValueSource<T>(default!);
    }
```?

Reply:
That won't work either. You'd need this:

```
public static ConstantValueSource<T> EmptyValueType<T>() where T : struct;
public static ConstantValueSource<T?> EmptyReferenceType<T>() where T : class;
```

There is no constraint available that allows one method to satisfy both usages while also disallowing non-nullable reference types.

Reply:
See also #38260

Reply:
I see.

Reply:
I'll follow up in a separate PR.